### DISCUSSION TITLE
Undo

### DISCUSSION THREAD

**Discussion by joerlane** [meerk40t/meerk40t#1319] [General]
Does undo need to be a feature before 0.8 release?

**Comment by tiger12506** [comment]
No, not unless it is implemented fully. I don't believe that losing classification of element ref nodes is acceptable.

**Comment by tatarize** [comment]
Well, that would be phase 1. The idea was get loading and restoring of tree states that *appears* to work. But, because of the fact that all nodes in the tree need to be copied. There are quickie values in nodes for `._references` that store things. So it saves restore states at certain points and *sometimes* would restore that state. But, that was going through all the nodes making copies. Then going through that node list and pointing copies of children to copies of parents to set the tree up correctly.

In theory it might be fixable fairly reasonably quickly but I could totally see some bugs cropping up. For example modifying things and altering things triggers the right controls but "adding" and "deleting" triggers nothing. So you can't currently undo an add or delete.

**Comment by tatarize** [comment]
If you alter something by making it larger or rotating it, that's a "modify" if you change the actual core objects that's "alteration". We have that stuff flagged everywhere things perform those actions. However, we don't have similar hooks for adding or removing. So those aren't currently marked to flag a undo point. Also, it currently lets you modify the undone state. So really it needs to swap out the state without every modifying those states. It currently copies the tree each time it's told to save_restore_state. But, then when you do undo, it switches that copy of the tree in for the current copy of the tree. That's a problem because it means you're not modifying that save_restore_state in the tree. And it sort of deleted the current tree state so redo doesn't quite work right. In theory if that stuff worked it would just be undo-redo that doesn't save adding/removing and loses all the reference nodes (currently) when that happens.

It's still not there, and it's really hard to know how deep those issues go until it's done.

**Comment by joerlane** [comment]
I vote to add an Edit menu to MeerK40t. With Undo, Cut, Copy, Paste, Paste Special, Delete, Select All; and maybe some Find features?

That's what I see "normally" in Mac applications. Check linux and windows and see what's normal cross-platform.
